File: backend/src/routes/wallet.js
import express from 'express';
import { v4 as uuidv4 } from 'uuid';
import { getMany, getOne, query } from '../db.js';
import { authMiddleware } from '../middleware/auth.js';

const router = express.Router();

// All routes require authentication
router.use(authMiddleware);

// Get wallet summary - cumulative savings by month
router.get('/summary', async (req, res) => {
  try {
    const { currency = 'RON' } = req.query;
    
    const sql = `
      SELECT 
        month,
        SUM(CASE WHEN type = 'income' THEN amount ELSE 0 END) as total_income,
        SUM(CASE WHEN type = 'expense' THEN amount ELSE 0 END) as total_expense,
        SUM(CASE WHEN type = 'income' THEN amount ELSE -amount END) as balance
      FROM transactions
      WHERE created_by = $1 AND currency = $2
      GROUP BY month
      ORDER BY month ASC
    `;
    
    const monthlyData = await getMany(sql, [req.user.id, currency]);
    
    // Calculate cumulative savings
    let cumulativeSavings = 0;
    const walletData = monthlyData.map(month => {
      const monthBalance = parseFloat(month.balance) || 0;
      cumulativeSavings += monthBalance;
      
      return {
        month: month.month,
        total_income: parseFloat(month.total_income) || 0,
        total_expense: parseFloat(month.total_expense) || 0,
        balance: monthBalance,
        cumulative_savings: cumulativeSavings
      };
    });
    
    res.json({
      monthly_data: walletData,
      total_cumulative: cumulativeSavings
    });
  } catch (error) {
    console.error('Get wallet summary error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Get current available balance across all currencies
router.get('/balance', async (req, res) => {
  try {
    const sql = `
      SELECT
        currency,
        SUM(CASE WHEN type = 'income' THEN amount ELSE -amount END) as available_balance
      FROM transactions
      WHERE created_by = $1
      GROUP BY currency
      ORDER BY currency
    `;

    const balances = await getMany(sql, [req.user.id]);

    // Calculate total in RON (need exchange rates)
    const ratesSql = 'SELECT currency, rate FROM exchange_rates';
    const rates = await getMany(ratesSql, []);

    const ratesMap = {};
    rates.forEach(r => ratesMap[r.currency] = parseFloat(r.rate));

    let totalInRon = 0;
    const balancesByCurrency = {};

    balances.forEach(b => {
      const balance = parseFloat(b.available_balance) || 0;
      balancesByCurrency[b.currency] = balance;

      if (b.currency === 'RON') {
        totalInRon += balance;
      } else if (ratesMap[b.currency]) {
        totalInRon += balance * ratesMap[b.currency];
      }
    });

    // Calculate meal voucher balance
    const mealVoucherSql = `
      SELECT
        COALESCE(SUM(CASE WHEN type = 'income' THEN amount ELSE 0 END), 0) as total_income,
        COALESCE(SUM(CASE WHEN type = 'expense' THEN amount ELSE 0 END), 0) as total_expense
      FROM transactions
      WHERE created_by = $1 AND is_meal_voucher = true
    `;
    const mealVoucherResult = await getOne(mealVoucherSql, [req.user.id]);
    const mealVoucherIncome = parseFloat(mealVoucherResult?.total_income) || 0;
    const mealVoucherExpense = parseFloat(mealVoucherResult?.total_expense) || 0;
    const mealVoucherBalance = mealVoucherIncome - mealVoucherExpense;

    res.json({
      balances: balancesByCurrency,
      total_in_ron: totalInRon,
      exchange_rates: ratesMap,
      meal_voucher_balance: mealVoucherBalance,
      meal_voucher_income: mealVoucherIncome,
      meal_voucher_expense: mealVoucherExpense
    });
  } catch (error) {
    console.error('Get wallet balance error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Get meal vouchers balance
router.get('/meal-vouchers', async (req, res) => {
  try {
    const sql = `
      SELECT
        COALESCE(SUM(CASE WHEN type = 'income' THEN amount ELSE 0 END), 0) as total_income,
        COALESCE(SUM(CASE WHEN type = 'expense' THEN amount ELSE 0 END), 0) as total_expense
      FROM transactions
      WHERE created_by = $1 AND is_meal_voucher = true
    `;

    const result = await getOne(sql, [req.user.id]);
    const totalIncome = parseFloat(result?.total_income) || 0;
    const totalExpense = parseFloat(result?.total_expense) || 0;
    const balance = totalIncome - totalExpense;

    res.json({
      balance: balance,
      total_income: totalIncome,
      total_expense: totalExpense
    });
  } catch (error) {
    console.error('Get meal vouchers balance error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// List meal voucher transactions
router.get('/meal-vouchers/transactions', async (req, res) => {
  try {
    const limit = Math.min(parseInt(req.query.limit) || 50, 500);
    const offset = parseInt(req.query.offset) || 0;

    const sql = `
      SELECT id, amount, type, category_id, category_name, description, date, month,
             currency, is_recurring, recurring_group_id, recurring_day, is_meal_voucher, created_by,
             created_date, updated_date
      FROM transactions
      WHERE created_by = $1 AND is_meal_voucher = true
      ORDER BY date DESC, created_date DESC
      LIMIT $2 OFFSET $3
    `;

    const transactions = await getMany(sql, [req.user.id, limit, offset]);
    res.json(transactions);
  } catch (error) {
    console.error('List meal voucher transactions error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Receive meal vouchers (income)
router.post('/meal-vouchers/receive', async (req, res) => {
  try {
    const { amount, description, transaction_date, date, is_recurring = true } = req.body;

    if (!amount || amount <= 0) {
      return res.status(400).json({ error: 'Amount must be greater than 0' });
    }

    const txDate = (date || transaction_date) ? new Date(date || transaction_date) : new Date();
    const month = `${txDate.getFullYear()}-${String(txDate.getMonth() + 1).padStart(2, '0')}`;
    const day = txDate.getDate();

    // Find "Bonuri de masă" category
    const category = await getOne(
      'SELECT id FROM budget_categories WHERE name = $1 AND created_by = $2',
      ['Bonuri de masă', req.user.id]
    );

    const categoryId = category?.id || null;
    const transactionId = uuidv4();
    const recurringGroupId = is_recurring ? uuidv4() : null;

    const sql = `
      INSERT INTO transactions (
        id, amount, type, category_id, category_name, description, date, month,
        currency, is_recurring, recurring_group_id, recurring_day, is_meal_voucher, created_by
      ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14)
      RETURNING *
    `;

    const result = await getOne(sql, [
      transactionId,
      amount,
      'income',
      categoryId,
      'Bonuri de masă',
      description || null,
      txDate.toISOString().split('T')[0],
      month,
      'RON',
      is_recurring,
      recurringGroupId,
      day,
      true,
      req.user.id
    ]);

    res.status(201).json(result);
  } catch (error) {
    console.error('Receive meal vouchers error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Spend meal vouchers (expense)
router.post('/meal-vouchers/spend', async (req, res) => {
  try {
    const { amount, description, transaction_date, date } = req.body;

    if (!amount || amount <= 0) {
      return res.status(400).json({ error: 'Amount must be greater than 0' });
    }

    // Check meal voucher balance
    const balanceSql = `
      SELECT
        COALESCE(SUM(CASE WHEN type = 'income' THEN amount ELSE 0 END), 0) as total_income,
        COALESCE(SUM(CASE WHEN type = 'expense' THEN amount ELSE 0 END), 0) as total_expense
      FROM transactions
      WHERE created_by = $1 AND is_meal_voucher = true
    `;
    const balanceResult = await getOne(balanceSql, [req.user.id]);
    const currentBalance = (parseFloat(balanceResult?.total_income) || 0) - (parseFloat(balanceResult?.total_expense) || 0);

    if (amount > currentBalance) {
      return res.status(400).json({
        error: `Sold insuficient de bonuri de masă. Disponibil: ${currentBalance.toFixed(2)} RON, Solicitat: ${amount} RON`
      });
    }

    const txDate = (date || transaction_date) ? new Date(date || transaction_date) : new Date();
    const month = `${txDate.getFullYear()}-${String(txDate.getMonth() + 1).padStart(2, '0')}`;

    // Find "Alimente" category
    const category = await getOne(
      'SELECT id FROM budget_categories WHERE name = $1 AND created_by = $2',
      ['Alimente', req.user.id]
    );

    const categoryId = category?.id || null;
    const transactionId = uuidv4();

    const sql = `
      INSERT INTO transactions (
        id, amount, type, category_id, category_name, description, date, month,
        currency, is_recurring, is_meal_voucher, created_by
      ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12)
      RETURNING *
    `;

    const result = await getOne(sql, [
      transactionId,
      amount,
      'expense',
      categoryId,
      'Alimente',
      description || null,
      txDate.toISOString().split('T')[0],
      month,
      'RON',
      false,
      true,
      req.user.id
    ]);

    res.status(201).json(result);
  } catch (error) {
    console.error('Spend meal vouchers error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Delete meal voucher transaction
router.delete('/meal-vouchers/transactions/:id', async (req, res) => {
  try {
    const { id } = req.params;

    const result = await getOne(
      'DELETE FROM transactions WHERE id = $1 AND created_by = $2 AND is_meal_voucher = true RETURNING id',
      [id, req.user.id]
    );

    if (!result) {
      return res.status(404).json({ error: 'Meal voucher transaction not found' });
    }

    res.json({ message: 'Meal voucher transaction deleted successfully', success: true });
  } catch (error) {
    console.error('Delete meal voucher transaction error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

export default router;

File: fastapi_app/routes/wallet.py
"""
Wallet and summary routes.
"""
import uuid
from datetime import date
from typing import List, Optional
from decimal import Decimal
from fastapi import APIRouter, Depends, HTTPException, status, Query
from models import (
    WalletSummaryResponse, WalletBalance, MonthlySummary, MealVoucherBalance,
    TransactionResponse, MessageResponse
)
from auth import get_current_user, User
from database import query, query_one, execute

router = APIRouter(prefix="/wallet", tags=["Wallet"])


@router.get("/balance", response_model=List[WalletBalance])
async def get_balance(current_user: User = Depends(get_current_user)):
    """
    Get current balance broken down by currency.

    Returns total income, expenses, and balance for each currency used.
    """
    balances = query(
        """
        SELECT
            currency,
            COALESCE(SUM(CASE WHEN type = 'income' THEN amount ELSE 0 END), 0) as total_income,
            COALESCE(SUM(CASE WHEN type = 'expense' THEN amount ELSE 0 END), 0) as total_expenses,
            COALESCE(SUM(CASE WHEN type = 'income' THEN amount ELSE -amount END), 0) as balance,
            COUNT(*) as transaction_count
        FROM transactions
        WHERE created_by = %s
        GROUP BY currency
        ORDER BY currency
        """,
        (current_user.id,)
    )

    return [
        WalletBalance(
            currency=b["currency"] or "RON",
            total_income=b["total_income"],
            total_expenses=b["total_expenses"],
            balance=b["balance"],
            transaction_count=b["transaction_count"]
        )
        for b in balances
    ]


@router.get("/summary", response_model=WalletSummaryResponse)
async def get_summary(current_user: User = Depends(get_current_user)):
    """
    Get comprehensive wallet summary including:
    - Balance by currency
    - Monthly summaries (income, expenses, net savings)
    - Cumulative savings over time
    """
    # Get balances by currency
    balances = query(
        """
        SELECT
            currency,
            COALESCE(SUM(CASE WHEN type = 'income' THEN amount ELSE 0 END), 0) as total_income,
            COALESCE(SUM(CASE WHEN type = 'expense' THEN amount ELSE 0 END), 0) as total_expenses,
            COALESCE(SUM(CASE WHEN type = 'income' THEN amount ELSE -amount END), 0) as balance,
            COUNT(*) as transaction_count
        FROM transactions
        WHERE created_by = %s
        GROUP BY currency
        ORDER BY currency
        """,
        (current_user.id,)
    )

    # Get exchange rates
    exchange_rates = query("SELECT currency, rate FROM exchange_rates")
    rates_map = {r["currency"]: r["rate"] for r in exchange_rates}

    # Calculate total balance in RON
    total_balance_ron = Decimal("0")
    wallet_balances = []
    for b in balances:
        currency = b["currency"] or "RON"
        balance = b["balance"]
        wallet_balances.append(WalletBalance(
            currency=currency,
            total_income=b["total_income"],
            total_expenses=b["total_expenses"],
            balance=balance,
            transaction_count=b["transaction_count"]
        ))
        # Convert to RON
        if currency == "RON":
            total_balance_ron += balance
        elif currency in rates_map:
            total_balance_ron += balance * rates_map[currency]

    # Get monthly summaries with cumulative savings
    monthly_data = query(
        """
        SELECT
            month,
            COALESCE(SUM(CASE WHEN type = 'income' THEN amount ELSE 0 END), 0) as total_income,
            COALESCE(SUM(CASE WHEN type = 'expense' THEN amount ELSE 0 END), 0) as total_expenses,
            COUNT(*) as transaction_count
        FROM transactions
        WHERE created_by = %s
        GROUP BY month
        ORDER BY month ASC
        """,
        (current_user.id,)
    )

    # Calculate cumulative savings
    cumulative = Decimal("0")
    monthly_summaries = []
    for m in monthly_data:
        net_savings = m["total_income"] - m["total_expenses"]
        cumulative += net_savings
        monthly_summaries.append(MonthlySummary(
            month=m["month"],
            total_income=m["total_income"],
            total_expenses=m["total_expenses"],
            net_savings=net_savings,
            cumulative_savings=cumulative,
            transaction_count=m["transaction_count"]
        ))

    return WalletSummaryResponse(
        balances=wallet_balances,
        monthly_summaries=monthly_summaries,
        total_balance_ron=total_balance_ron
    )


@router.get("/monthly/{month}", response_model=MonthlySummary)
async def get_monthly_summary(
    month: str,
    current_user: User = Depends(get_current_user)
):
    """
    Get summary for a specific month.

    - **month**: Month in YYYY-MM format
    """
    data = query(
        """
        SELECT
            month,
            COALESCE(SUM(CASE WHEN type = 'income' THEN amount ELSE 0 END), 0) as total_income,
            COALESCE(SUM(CASE WHEN type = 'expense' THEN amount ELSE 0 END), 0) as total_expenses,
            COUNT(*) as transaction_count
        FROM transactions
        WHERE created_by = %s AND month = %s
        GROUP BY month
        """,
        (current_user.id, month)
    )

    if not data:
        return MonthlySummary(
            month=month,
            total_income=Decimal("0"),
            total_expenses=Decimal("0"),
            net_savings=Decimal("0"),
            cumulative_savings=Decimal("0"),
            transaction_count=0
        )

    m = data[0]
    net_savings = m["total_income"] - m["total_expenses"]

    # Calculate cumulative savings up to this month
    cumulative = query(
        """
        SELECT COALESCE(SUM(
            CASE WHEN type = 'income' THEN amount ELSE -amount END
        ), 0) as cumulative
        FROM transactions
        WHERE created_by = %s AND month <= %s
        """,
        (current_user.id, month)
    )

    return MonthlySummary(
        month=m["month"],
        total_income=m["total_income"],
        total_expenses=m["total_expenses"],
        net_savings=net_savings,
        cumulative_savings=cumulative[0]["cumulative"] if cumulative else Decimal("0"),
        transaction_count=m["transaction_count"]
    )


@router.get("/stats/categories")
async def get_category_stats(
    current_user: User = Depends(get_current_user)
):
    """
    Get spending/income breakdown by category.

    Returns statistics grouped by category with totals and counts.
    """
    stats = query(
        """
        SELECT
            c.id as category_id,
            c.name as category_name,
            c.type,
            c.icon,
            c.color,
            COUNT(t.id) as transaction_count,
            SUM(t.amount) as total_amount,
            AVG(t.amount) as average_amount,
            MIN(t.date) as first_transaction,
            MAX(t.date) as last_transaction
        FROM budget_categories c
        LEFT JOIN transactions t ON c.id = t.category_id AND t.created_by = %s
        WHERE c.created_by = %s
        GROUP BY c.id, c.name, c.type, c.icon, c.color
        ORDER BY c.type, total_amount DESC NULLS LAST
        """,
        (current_user.id, current_user.id)
    )

    return stats


@router.get("/meal-vouchers", response_model=MealVoucherBalance)
async def get_meal_voucher_balance(current_user: User = Depends(get_current_user)):
    """
    Get meal voucher balance.

    Returns total meal voucher income, expenses, and remaining balance.
    Meal vouchers can only be spent on food category (Alimente).
    """
    result = query(
        """
        SELECT
            COALESCE(SUM(CASE WHEN type = 'income' THEN amount ELSE 0 END), 0) as total_income,
            COALESCE(SUM(CASE WHEN type = 'expense' THEN amount ELSE 0 END), 0) as total_expense
        FROM transactions
        WHERE created_by = %s AND is_meal_voucher = true
        """,
        (current_user.id,)
    )

    if not result:
        return MealVoucherBalance(
            balance=Decimal("0"),
            total_income=Decimal("0"),
            total_expense=Decimal("0")
        )

    total_income = Decimal(str(result[0]["total_income"]))
    total_expense = Decimal(str(result[0]["total_expense"]))
    balance = total_income - total_expense

    return MealVoucherBalance(
        balance=balance,
        total_income=total_income,
        total_expense=total_expense
    )


@router.post("/meal-vouchers/receive", response_model=TransactionResponse, status_code=status.HTTP_201_CREATED)
async def receive_meal_vouchers(
    amount: Decimal,
    description: Optional[str] = None,
    transaction_date: Optional[date] = None,
    is_recurring: bool = True,
    current_user: User = Depends(get_current_user)
):
    """
    Receive meal vouchers (income).

    Automatically sets is_meal_voucher=true and currency=RON.
    Defaults is_recurring=true since meal vouchers come monthly.

    - **amount**: Amount of meal vouchers received (in RON)
    - **description**: Optional description (e.g., "Bonuri de masă Februarie 2026")
    - **transaction_date**: Date of receipt (defaults to today)
    - **is_recurring**: Whether this is a recurring monthly income (default: true)
    """
    # Use today's date if not provided
    if transaction_date is None:
        transaction_date = date.today()

    # Calculate month from date
    month = transaction_date.strftime("%Y-%m")

    # Find or create "Bonuri de masă" category
    category = query_one(
        "SELECT id FROM budget_categories WHERE name = %s AND created_by = %s",
        ("Bonuri de masă", current_user.id)
    )

    category_id = category["id"] if category else None
    transaction_id = str(uuid.uuid4())

    # Generate recurring_group_id for recurring transactions
    recurring_group_id = None
    if is_recurring:
        recurring_group_id = str(uuid.uuid4())

    result = execute(
        """
        INSERT INTO transactions (
            id, amount, type, category_id, category_name, description, date, month,
            currency, is_recurring, recurring_group_id, recurring_day, is_meal_voucher, created_by
        ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
        RETURNING *
        """,
        (
            transaction_id,
            amount,
            "income",
            category_id,
            "Bonuri de masă",
            description,
            transaction_date,
            month,
            "RON",
            is_recurring,
            recurring_group_id,
            transaction_date.day,
            True,  # is_meal_voucher
            current_user.id
        )
    )

    return TransactionResponse(**result)


@router.post("/meal-vouchers/spend", response_model=TransactionResponse, status_code=status.HTTP_201_CREATED)
async def spend_meal_vouchers(
    amount: Decimal,
    description: Optional[str] = None,
    transaction_date: Optional[date] = None,
    current_user: User = Depends(get_current_user)
):
    """
    Spend meal vouchers on food (Alimente category only).

    Automatically sets category_name="Alimente", is_meal_voucher=true, currency=RON.
    Validates that sufficient meal voucher balance is available.

    - **amount**: Amount to spend (in RON)
    - **description**: Optional description (e.g., "Prânz restaurant")
    - **transaction_date**: Date of spending (defaults to today)
    """
    # Use today's date if not provided
    if transaction_date is None:
        transaction_date = date.today()

    # Calculate month from date
    month = transaction_date.strftime("%Y-%m")

    # Check meal voucher balance
    balance_result = query(
        """
        SELECT
            COALESCE(SUM(CASE WHEN type = 'income' THEN amount ELSE 0 END), 0) as total_income,
            COALESCE(SUM(CASE WHEN type = 'expense' THEN amount ELSE 0 END), 0) as total_expense
        FROM transactions
        WHERE created_by = %s AND is_meal_voucher = true
        """,
        (current_user.id,)
    )

    total_income = Decimal(str(balance_result[0]["total_income"])) if balance_result else Decimal("0")
    total_expense = Decimal(str(balance_result[0]["total_expense"])) if balance_result else Decimal("0")
    current_balance = total_income - total_expense

    if amount > current_balance:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Insufficient meal voucher balance. Available: {current_balance} RON, Requested: {amount} RON"
        )

    # Find or create "Alimente" category
    category = query_one(
        "SELECT id FROM budget_categories WHERE name = %s AND created_by = %s",
        ("Alimente", current_user.id)
    )

    category_id = category["id"] if category else None
    transaction_id = str(uuid.uuid4())

    result = execute(
        """
        INSERT INTO transactions (
            id, amount, type, category_id, category_name, description, date, month,
            currency, is_recurring, is_meal_voucher, created_by
        ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
        RETURNING *
        """,
        (
            transaction_id,
            amount,
            "expense",
            category_id,
            "Alimente",
            description,
            transaction_date,
            month,
            "RON",
            False,  # is_recurring - expenses are not recurring
            True,   # is_meal_voucher
            current_user.id
        )
    )

    return TransactionResponse(**result)


@router.get("/meal-vouchers/transactions", response_model=List[TransactionResponse])
async def list_meal_voucher_transactions(
    limit: int = Query(50, ge=1, le=500),
    offset: int = Query(0, ge=0),
    current_user: User = Depends(get_current_user)
):
    """
    List all meal voucher transactions.

    Returns transactions where is_meal_voucher=true, ordered by date descending.

    - **limit**: Maximum number of results (default: 50, max: 500)
    - **offset**: Number of results to skip (default: 0)
    """
    transactions = query(
        """
        SELECT id, amount, type, category_id, category_name, description, date, month,
               currency, is_recurring, recurring_group_id, recurring_day, is_meal_voucher, created_by,
               created_date, updated_date
        FROM transactions
        WHERE created_by = %s AND is_meal_voucher = true
        ORDER BY date DESC, created_date DESC
        LIMIT %s OFFSET %s
        """,
        (current_user.id, limit, offset)
    )

    return [TransactionResponse(**t) for t in transactions]


@router.delete("/meal-vouchers/transactions/{transaction_id}", response_model=MessageResponse)
async def delete_meal_voucher_transaction(
    transaction_id: str,
    current_user: User = Depends(get_current_user)
):
    """
    Delete a meal voucher transaction.

    Only transactions belonging to the authenticated user and marked as meal vouchers can be deleted.
    """
    result = execute(
        "DELETE FROM transactions WHERE id = %s AND created_by = %s AND is_meal_voucher = true RETURNING id",
        (transaction_id, current_user.id)
    )

    if not result:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Meal voucher transaction not found"
        )

    return MessageResponse(message="Meal voucher transaction deleted successfully")